lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/files/JobJarMaker.java

```diff
@@ -96,55 +102,88 @@ private static File toFile(final FileResource fileProto) {
+      return this;
+    }
 
-    // Create a JAR File for the submission
-    final File jarFile = File.createTempFile(this.fileNames.getJobFolderPrefix(), this.fileNames.getJarFileSuffix());
+    public JarBuilder withGlobalFileSet(final Set<FileResource> withGlobalFiles) {
+      this.globalFiles = withGlobalFiles;
+      return this;
+    }
 
-    LOG.log(Level.FINE, "Creating job submission jar file: {0}", jarFile);
-    new JARFileMaker(jarFile).addChildren(jobSubmissionFolder).close();
+    public JarBuilder withConfiguration(final Configuration withConfiguration) {
+      this.configuration = withConfiguration;
+      return this;
+    }
 
-    if (this.deleteTempFilesOnExit) {
-      LOG.log(Level.FINE,
-          "Deleting the temporary job folder [{0}] and marking the jar file [{1}] for deletion after the JVM exits.",
-          new Object[]{jobSubmissionFolder.getAbsolutePath(), jarFile.getAbsolutePath()});
-      if (!jobSubmissionFolder.delete()) {
-        LOG.log(Level.WARNING, "Failed to delete [{0}]", jobSubmissionFolder.getAbsolutePath());
-      }
-      jarFile.deleteOnExit();
-    } else {
-      LOG.log(Level.FINE, "Keeping the temporary job folder [{0}] and jar file [{1}] available after job submission.",
-          new Object[]{jobSubmissionFolder.getAbsolutePath(), jarFile.getAbsolutePath()});
+    public JarBuilder withConfigurationFileName(final String withConfigurationFilename) {
+      this.configurationFilename = withConfigurationFilename;
+      return this;
     }
-    return jarFile;
-  }
 
-  private File makejobSubmissionFolder() throws IOException {
-    return Files.createTempDirectory(this.fileNames.getJobFolderPrefix()).toFile();
+    public File build() throws IOException {
+      // Copy all files to a local job submission folder
+      final File jobSubmissionFolder = makejobSubmissionFolder();
+      LOG.log(Level.FINE, "Staging submission in {0}", jobSubmissionFolder);
+
+      final File localFolder = new File(jobSubmissionFolder, JobJarMaker.this.fileNames.getLocalFolderName());
+      final File globalFolder = new File(jobSubmissionFolder, JobJarMaker.this.fileNames.getGlobalFolderName());
+
+      JobJarMaker.copy(this.globalFiles, globalFolder);
+      JobJarMaker.copy(this.localFiles, localFolder);
+
+      // Store the Configuration in the JAR file.
+      if (configuration != null && StringUtils.isNotBlank(this.configurationFilename)) {
+        JobJarMaker.this.configurationSerializer
+            .toFile(configuration, new File(localFolder, this.configurationFilename));
+      }
+
+      // Create a JAR File for the submission
+      final File jarFile = File.createTempFile(JobJarMaker.this.fileNames.getJobFolderPrefix(),
+          JobJarMaker.this.fileNames.getJarFileSuffix());
+
+      LOG.log(Level.FINE, "Creating job submission jar file: {0}", jarFile);
+      new JARFileMaker(jarFile).addChildren(jobSubmissionFolder).close();
+
+      if (JobJarMaker.this.deleteTempFilesOnExit) {
```

Comment:
I think this Boolean should be moved into the Builder class. It is only used here.
After that the builder class could change to static class, as it does not need to hold a reference to outer class. You will not need to new 

Is this because of that Boolean is injected? Probably we could consider pass it in with builder pattern.

Reply:
Not sure what you mean by "After that the builder class could change to static class, as it does not need to hold a reference to outer class".
Builder pattern is there to simplify setting a large number of parameters and should serve only as a way for client code to set object properties. Application-wide settings (i.e. whether to delete temp files) should be set outside of builder.

Reply:
We discussed offline. For other's reference: https://stackoverflow.com/questions/16147405/whats-the-advantage-of-making-an-inner-class-as-static-with-java